Raise keyboard events using special keys

duracellko opened the pull request:
## Pull request description
Implements #101.
Added `KeyDown` and `KeyUp` helper methods to raise keyboard events with special keys. Available special keys are represented by `Key` enum. Special keys can be combined with control keys: `Key.Enter | Key.Alt`. `KeyPress` helper method is not added, because this event does not emit special keys.

`Key` enum is subset of values from Selenium `Keys` class. Implementation of `KeyEventFactory` is based on real values in Chrome browser. I created small component that displays KeyboardEventArgs values. Then simple Selenium application collected the values for all special keys.

Documentation has not been updated. Documentation of API reference should be enough.

### PR meta checklist
- [x] Pull request is targeting at `DEV` branch.
- [x] Pull request is linked to all related issues, if any.
- [x] I have read the _CONTRIBUTING.md_ document.

### Content checklist
- [x] My code follows the code style of this project and AspNetCore coding guidelines.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [x] I have added, updated or removed tests to according to my changes.
  - [x] All tests passed.

egil commented:
Thanks for this @duracellko. I hope I get some time to day to review and provide feedback. Might first be tomorrow though.

duracellko commented:
Open question: Current implementation does not return Task to wait for event dispatcher. Should this be changed to Async method?

egil reviewed (commented):
Hi @duracellko 

Wow, this turned out more complex than I expected :)

This is solid work as always from you, do see some suggestions in the review comments around the files.

I do have some concerns with the current implementation, that I think we need to address:

1. I do not see a way I can pass in a regular key, e.g. "A", nither by itself or in a combination with a control key, such as `CTRL+A`.
2. We need to include Mac keyboard keys as well (isnt it called the option key?).
3. I am not certain of the intuitiveness of using the `|` to indicate multiple pressed keys. To me, the `|` operator indicates "or" operation. A more correct operator would be the `&` "and" operator. But I actually think it is more intuitive if we simple allow users to pass `param Key[] keys` in. That would look like this: `KeyDown(Key.Enter, Key.Control)`.   
   
   It would be even nicer if it was possible pass `CTRL+A` like this: `KeyDown(Key.Control, "a")`. I think that should be possible if `Key` was a struct/class instead of an enum. Each of the control characters would be a static property on it, and then it would also need an implicit conversion from string, making it possible to both pass `Key` and a string to the same `param Key[] keys` list, as in the example.

What do you think about that?

egil commented:
> Open question: Current implementation does not return Task to wait for event dispatcher. Should this be changed to Async method?

Short answer, no.

Longer answer: I am actually thinking about removing the "async" versions of the event dispatcher helper methods, as they are not useful at all as far as I can see, and just a leftover from when I imported some of Steves testing prototype code, and didn't understand what was going on. 

If the event handler is async and is not done processing when the dispatcher returns, it is most likely waiting for data from the user, e.g. the test code, so awaiting it there does nothing. Removing the async methods would simply makes things less confusing to folks.

duracellko commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,257 @@
+		/// <summary>
+		/// No keyboard key pressed. This option raises no event.
+		/// </summary>
+		Null = 0,
```
I agree `None` is better name.

I don't think this value will be used by consumer. I just prefer for `enum` to have defined default value. In this case: `default(Key)` should have defined value and not just `0`.

duracellko commented on src/bunit.web/EventDispatchExtensions/KeyboardEventDispatchExtensions.cs:
```diff
@@ -43,6 +43,20 @@ public static void KeyDown(this IElement element, string key, string? code = def
+		/// <param name="key">The keyboard key to raise the event for.</param>
+		public static void KeyDown(this IElement element, Key key)
+		{
+			var eventArgs = KeysEventFactory.CreateKeyboardEventArgs(key);
+			if (eventArgs != null)
+			{
+				KeyDown(element, eventArgs);
+			}
```
That is good idea. Originally I just followed Selenium design. Selenium does not do anything, when `Null` is provided.

One advantage of not throwing exception is that data driven tests can cover cases with pressed and not pressed key. But that can be explicitly expressed in the test itself.

duracellko commented:
> I do have some concerns with the current implementation, that I think we need to address:
> 
> 1. I do not see a way I can pass in a regular key, e.g. "A", nither by itself or in a combination with a control key, such as `CTRL+A`.

With regular keys, it's hard to tell, where to stop. Letters and numbers? What special characters? Currency symbols? Non-english letters? Maybe it can be cleared up after point 3.

> 2. We need to include Mac keyboard keys as well (isnt it called the option key?).

I don't have Mac, I cannot verify, how Option key behaves. Isn't it emulated as Alt key?

> 3. I am not certain of the intuitiveness of using the `|` to indicate multiple pressed keys. To me, the `|` operator indicates "or" operation. A more correct operator would be the `&` "and" operator. But I actually think it is more intuitive if we simple allow users to pass `param Key[] keys` in. That would look like this: `KeyDown(Key.Enter, Key.Control)`.
>    It would be even nicer if it was possible pass `CTRL+A` like this: `KeyDown(Key.Control, "a")`. I think that should be possible if `Key` was a struct/class instead of an enum. Each of the control characters would be a static property on it, and then it would also need an implicit conversion from string, making it possible to both pass `Key` and a string to the same `param Key[] keys` list, as in the example.
> 
> What do you think about that?

I think separating of input to key and key-modifier might be good idea. It removes that weird key-bit arithmetic that works. It is a bit weird, because it is possible to do `Key.Enter | Key.Escape`, but it's not valid option.
Separating key and key modifiers also enables more cases to simulate. For example, pressing Control, while holding Shift, is different than pressing Shift, while holding Control. I realized it's not possible to express both cases with current implementation.

I see following option:

**Option 1**

Use enums for key and key modifiers.

```csharp
// Represents key being pressed or released
public enum Key
{
  Enter,
  Escape,
  ...
  Control,
  Alt
}

// Represents key being hold down, when event occured.
[Flags]
public enum KeyModifiers
{
  Control,
  Shift,
  Alt
}

public void KeyDown(this IElement element, Key key, KeyModifiers keyModifiers)

// Example:
element.KeyDown(Key.Enter, KeyModifiers.Control | KeyModifiers.Shift);
```

**Option 2**

Use struct for KeyModifiers.

```csharp
public struct KeyModifiers
{
  public KeyModifiers(bool control, bool alt, bool shift)
  {
    Control = control;
    Alt = alt;
    Shift = shift;
  }
  public bool Control { get; }
  public bool Alt { get; }
  public bool Shift { get; }

  // Implement Equals and GetHashCode
}

// Same Key type as in Option 1
public void KeyDown(this IElement element, Key key, KeyModifiers keyModifiers)

// Example:
element.KeyDown(Key.Enter, new KeyModifiers(true, false, true));
```

**Option 3**

Use parameters. It's not much different from the generic `KeyDown` method.

```csharp
// Same Key type as in Option 1
public void KeyDown(this IElement element, Key key, bool controlKey = false, bool altKey = false, bool shiftKey = false)

// Example:
element.KeyDown(Key.Enter, controlKey:true, shiftKey: true);
```

**Remarks**

I wouldn't go with array as input. Consumer might do `element.KeyDown(Key.Enter, Key.Escape)`, but that does not make sense. Therefore explicit splitting key and key modifiers is probably better way.

duracellko commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,257 @@
+	/// <summary>
+	/// Special keys that can be dispatched in keyboard events.
+	/// </summary>
+	public enum Key
```
Actually I omitted `[Flags]` attribute, because it's not really flags enum. bit logic can be used only on control keys `Key.Control | Key.Shift`, but not on other keys `Key.Enter | Key.Escape`.

egil commented:
I like your thought about trying to make something work that the compiler will complain about. But I want to consider just doing runtime validation if it means providing a simpler API:

Something like this could work, with the runtime of the valid combinations:

```
using Microsoft.AspNetCore.Components.Web;

namespace KeyNameSpace
{
    public class KeyCombination
    {
        // This class contains validation logic in ctor that checks
        // if it (the key combination) is legal.

        public static KeyCombination operator +(KeyCombination a, Key b) => new KeyCombination();
        public static implicit operator KeyboardEventArgs(KeyCombination key) => new KeyboardEventArgs();
    }

    public class Key
    {
        public string Value { get; }
        public string Code { get; }

        private Key(string key, string code)
        {
            Value = key;
            Code = code;
        }

        public static Key Enter { get; } = new Key("Enter", "Enter");
        public static Key Control { get; } = new Key("Control", "ControlLeft");
        // ... add the rest of the special keys here

        public static implicit operator Key(char key) => new Key(key.ToString(), key.ToString());
        public static KeyCombination operator +(Key a, Key b) => new KeyCombination();
        public static implicit operator KeyboardEventArgs(Key key) => new KeyboardEventArgs();
    }
}

namespace UserNameSpace
{
    using KeyNameSpace;
    using static KeyNameSpace.Key;

    public class Dummy
    {
        public static void KeyUp(KeyboardEventArgs args) {}

        public static void Blah()
        {
            KeyUp(Control + Enter);
            KeyUp(Control + Enter + 'x');
            KeyUp(Control + 'x');
            KeyUp(Enter);
            KeyUp(Enter);
        }
    }
}
```

I like this because:

- The plus operator makes it obvious that it is a combination that is being provided.
- The `using static` trick, which makes the code very expressive I think. 
- This also has the added benefit that we do not need to add new methods to the keyboard event  event dispatcher helpers, as the implicit conversion to KeyboardEventArgs  does the work for us.
- Use of chars to ensure folks dont send in more than one char (or is that incompatible with certain languages?)

Obviously more properties could needs to be added to the Key and KeyCombination class. Its just stubs.

What do you think?

duracellko commented:
Good idea. I will try to update PR later.

I assume that `Control + Enter + 'x'` is supposed to throw exception.

egil commented:
> Good idea. I will try to update PR later.
> 
> I assume that `Control + Enter + 'x'` is supposed to throw exception.

Only if it is an invalid key combination. It was just an example of a combination, did not think about its validity or not :)

duracellko commented:
I created new `Key` class. Let me know, what you think.
Tests are still in progress.

egil reviewed (commented):
Looks good so far. I see a few places where the indention is not right, but a ctrl+k+d should fix that.

duracellko commented:
I finished the implementation and testing.

egil commented:
> I finished the implementation and testing.

Sorry for the radio silence. Been super busy last few days. Will take a look asap this weekend and provide feedback.

egil reviewed (changes requested):
This is very good. I have a few additional things I think is worth fixing, but this close to done.

Suggestions:

- I know I said this could be done without modifying `KeyboardEventDispatchExtensions`, but lets add a new overload similar to `KeyUp(this IElement element, string key, string? code = default, float location = default, bool repeat = default, bool ctrlKey = default, bool shiftKey = default, bool altKey = default, bool metaKey = default, string? type = default)`, just with key and code replaced by `Key`.
- I think `new Key("Enter")` should return an `Key.Enter`, and similar for the other special keys. This might require you to drop ctors and go with factory methods instead. 
That will make this key passing method more interchangeable with an older version. 

Thanks!

duracellko commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,550 @@
+
+			static bool IsSuitableForCombination(Key key)
+			{
+				return (key.Value == Control.Value && key.Code == Control.Code) ||
+					(key.Value == Shift.Value && key.Code == Shift.Code) ||
+					(key.Value == Alt.Value && key.Code == Alt.Code) ||
+					(key.Value == Command.Value && key.Code == Command.Code);
```
Do you mean to rename `Key.Command` to `Key.Meta`? Because that is code/value of Command key. This was taken from Selenium. I sent Command key from selenium and browser displayed Meta value.

egil replied:
I just mean that `Key.Meta` should be combinable like `Key.Control` is.

duracellko replied:
But there is no `Key.Meta`. It is named `Key.Command`.

egil replied:
Ahh ok. Nevermind then 😂

duracellko commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,550 @@
+				return true;
+			}
+
+			return !(x is null) && x.Equals(y);
```
I even missed this is in C#. 😊

duracellko commented on tests/bunit.web.tests/EventDispatchExtensions/KeyboardEventDispatchExtensionsTest.cs:
```diff
@@ -29,9 +30,39 @@ public async Task CanRaiseEvents(MethodInfo helper)
+			var spy = CreateTriggerSpy(ElementName, "onkeyup");
+			spy.Trigger(element =>
+			{
+				element.KeyUp((Key)'A');
```
This will be possible, when we add new overload accepting `Key`.

egil replied:
Let's do that. See my general comment.

duracellko commented on tests/bunit.web.tests/EventDispatchExtensions/KeyTest.cs:
```diff
@@ -0,0 +1,479 @@
+		[InlineData("0", "0")]
+		[InlineData("&", "&")]
+		[InlineData("Key", "Key")]
+		[InlineData("TEST_test$44", "TEST_test$44")]
```
I don't want to speculate too much about key validation. Restricting it may cause problems in testing Chinese languages.
It's probably like validating email. It cannot be done right.

duracellko commented:
Comments are addressed. Also conflicts resolved. Hope I didn't miss anything.

duracellko commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,550 @@
+		[SuppressMessage("Usage", "CA2225:Operator overloads have named alternates", Justification = "Alternative method is named Combine")]
+		public static Key operator +(Key x, Key? y)
+		{
+			if (!(x is null))
```
This seems to make the build failed. `x is not null` is not recognized as checking for `null`. Should I revert the syntax or add exceptions to warnings?

duracellko replied:
Just checked, the warning should be fixed in next version of FxCop. https://github.com/dotnet/roslyn-analyzers/pull/4060

egil commented on src/bunit.web/EventDispatchExtensions/KeyboardEventDispatchExtensions.cs:
```diff
@@ -43,6 +43,14 @@ public static void KeyDown(this IElement element, string key, string? code = def
+		/// </summary>
+		/// <param name="element">The element to raise the event on.</param>
+		/// <param name="key">The keyboard key to pass to the event handler.</param>
+		public static void KeyDown(this IElement element, Key key) => _ = KeyDownAsync(element, key);
```
Compared to this: 

```
public static void KeyDown(this IElement element, string key, string? code = default, float location = default, bool repeat = default, bool ctrlKey = default, bool shiftKey = default, bool altKey = default, bool metaKey = default, string? type = default)
			=> KeyDownAsync(element, new KeyboardEventArgs { Key = key, Code = code!, Location = location, Repeat = repeat, CtrlKey = ctrlKey, ShiftKey = shiftKey, AltKey = altKey, MetaKey = metaKey, Type = type! });
```

There are a few arguments missing I think. Perhaps this is a better overload, that allows folks to do a little extra with the new Key type.

```suggestion
		public static void KeyDown(this IElement element, Key key, bool repeat = default, string? type = default) => ...   
```

What do you think (same for the two other KeyXXXX methods).

egil commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,617 @@
+		/// <param name="key">The character to convert to Key instance.</param>
+		/// <returns>The Key instance with character value.</returns>
+		[SuppressMessage("Usage", "CA2225:Operator overloads have named alternates", Justification = "Alternative method is named " + nameof(Get))]
+		public static implicit operator Key(char key) => Key.Get(key);
```
I think we need to add a `public static implicit operator Key(string key) => Key.Get(key);` as well to make it easy for folks to just get things working.

egil replied:
By just getting this to work I mean not having to use Key.Get for strings.

egil commented on src/bunit.web/EventDispatchExtensions/Key.cs:
```diff
@@ -0,0 +1,617 @@
+		/// </summary>
+		/// <param name="value">The key value.</param>
+		/// <param name="code">The key code of physical key.</param>
+		public static Key Get(string value, string code)
```
I am not sure the `Get` methods needs to be public. Unless there is a good reason, exposing multiple ways of doing the same thing, in this case the `Get` and implicit conversions from `char` and `string`, is more confusing to users. 

(Ive learned that the hard way with multiple ways to pass parameters to components...)

What is your thoughts around having `Get` methods public?

duracellko replied:
It is recommended practice to expose methods next to user-defined converters. There is even FxCop rule for that. I had to suppress it because I used different method names. But I would still follow the idea of the rule to have alternative methods. The main reason is that the library can be used with other languages than just C#.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2225

egil reviewed (changes requested):
Almost there, good work 👍 

You need to build the bUnit locally in release mode and fix the build issues. In release mode all warnings are treated as errors, and we cant merge with build errors.

Other than that, I only have two more suggestions, adding an extra implicit conversion from string to key, and consider making the `Get` methods private.

duracellko commented on CHANGELOG.md:
```diff
@@ -14,6 +14,20 @@ List of new features.
+element.KeyDown(Key.Enter + Key.Control); // Triggers onkeydown event with Ctrl + Enter
+element.KeyUp(Key.Control + Key.Shift + 'B'); // Triggers onkeyup event with Ctrl + Shift + B
+element.KeyPress('1'); // Triggers onkeypress event with key 1
+element.KeyDown(Key.Alt + Key.Get("<")); // Triggers onkeydown event with Alt + <
```
Example that string can be used too.

duracellko commented:
> Almost there, good work 👍
> 
> You need to build the bUnit locally in release mode and fix the build issues. In release mode all warnings are treated as errors, and we cant merge with build errors.
> 
> Other than that, I only have two more suggestions, adding an extra implicit conversion from string to key, and consider making the `Get` methods private.
> 
> Ps. my build script is bit messed up right now, but these three checks should pass:
> 
> ![image](https://user-images.githubusercontent.com/105649/95676758-23573700-0bb0-11eb-93ce-a7129e1f7bf2.png)

I commented on the problem with build above:

`x is not null` is not recognized as checking for null and it causes code analysis warning. Should I revert the syntax back to `!(x is null)` or add exceptions to warnings?

Just checked, the warning should be fixed in next version of FxCop. https://github.com/dotnet/roslyn-analyzers/pull/4060

egil commented:
> I commented on the problem with build above:

My bad. Didnt see that.
 
> `x is not null` is not recognized as checking for null and it causes code analysis warning. Should I revert the syntax back to `!(x is null)` or add exceptions to warnings?
> 
> Just checked, the warning should be fixed in next version of FxCop. [dotnet/roslyn-analyzers#4060](https://github.com/dotnet/roslyn-analyzers/pull/4060)

Leave the warnings in there for now, I will investigate and apply a fix locally.

duracellko commented:
Update:

1. Added `repeat` and `type` parameters to `KeyXXX` methods.
2. Added conversion from string to `Key`.
3. Added suppress warning to `+` operator.

My decision is to keep `Get` method public

egil reviewed (approved):
Thanks, looks good!
